casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -397,6 +399,36 @@ object Validate {
     } yield status
   }
 
+  /*
+   * This check must come before Validate.parents
+   */
+  def justificationFollows[F[_]: Monad: Log: BlockStore](
+      b: BlockMessage,
+      genesis: BlockMessage,
+      dag: BlockDag): F[Either[InvalidBlock, ValidBlock]] =
+    for {
+      latestMessages <- ProtoUtil.toLatestMessage[F](b.justifications, dag)
+      validators     = latestMessages.keySet
+      creatorLatestBlock <- ProtoUtil.creatorJustification(b).foldM(genesis) {
+                             case (_, Justification(_, latestBlockHash)) =>
+                               for {
+                                 latestBlock <- ProtoUtil.unsafeGetBlock[F](latestBlockHash)
+                               } yield latestBlock
+                           }
+      creatorBonds = ProtoUtil.bonds(creatorLatestBlock).map(_.validator).toSet
+      result       = creatorBonds == validators
+      status <- if (result) {
+                 Applicative[F].pure(Right(Valid))
+               } else {
+                 for {
+                   _ <- Log[F].warn(
+                         ignore(
+                           b,
+                           "the justifications of block are not follow from bonds of creator justification block"))
```

Comment:
the justification of the block does not follow from ... 

Also can you mkstring from the list of justifications and print them out with PrettyPrinter.buildString(...) ?

Reply:
Done